app/utils/ContextErrorMaping.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2019 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package utils
+
+import play.api.data.{FormError, Mapping}
+import play.api.data.validation.Constraint
+
+case class ContextErrorMapping[T](origin: Mapping[T]) extends Mapping[T] {
+  override val key: String = origin.key
+  override val mappings: Seq[Mapping[_]] = origin.mappings
+  override val constraints: Seq[Constraint[T]] = origin.constraints
+
+  private def fixErrors(errors: Seq[FormError]): Seq[FormError] =
+    errors.map {
+      case formError if formError.key.nonEmpty =>
+        formError
+      case formError =>
+        origin.mappings.collectFirst {
+          case mappingForTest if mappingForTest.key.nonEmpty && formError.message.startsWith(mappingForTest.key) =>
+            formError.copy(mappingForTest.key)
+        }.getOrElse(formError)
+    }
```

Comment:
What this method is doing?

Reply:
"Magic" Patryk, magic.

Reply:
Checks if any of global `FormError` message starts with any of `name` used in form and then if start move that global error to field with that name

Reply:
Basically, it's looking for FormErrors that don't have a key (i.e. they are not linked to a field on the form) and then adds a key based on the first part of the message matching the id of a field on the form.   So if we have an error like this - FormError("", "ducr.error") and the form contains a field with the id "ducr" then the error gets transformed to FormError("ducr", "ducr.error").

This is all done so that when the error is rendered the field in question gets highlighted.

The only slight downside of this implementation that I can see is that it relies on the error messages starting with the id of the field to which they relate.

Reply:
Thank you, yeah, understand now